Add vector value linking

KoBeWi opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/144

![](https://user-images.githubusercontent.com/2223172/158083370-56343a50-52b5-4901-97d6-ce1312469379.gif)
![](https://user-images.githubusercontent.com/2223172/161657524-c66160b4-d42b-49b7-9bdb-5edf6bf82473.gif)
Linked values change proportionally (unless any of them is 0). The link button only appears if property has `PROPERTY_HINT_LINK` (for now only Node2D/3D/Control `scale`).

The linked/unlinked icon is a placeholder, I did my best :/ I'll optimize them when they are replaced.

On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.

fire-forge commented:
This looks great! However, I think the link button should be moved to the right side as it'd make more sense there than on the left.

Also, are you considering doing this for Vector3 as well? The same argument about scale usually being the same on all axes can be applied to 3D too.

KoBeWi commented:
> Also, are you considering doing this for Vector3 as well?

Not in this PR. I didn't want it too big and we still need to agree on the design of this feature.

akien-mga commented:
> On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.

This is actually possible already, see `ADD_LINKED_PROPERTY` in `object.h`.

groud commented:
As discussed on a the PR meeting, the PR is very good but would need small changes:
- The link button should be hidden when the hint is not here. Later on, if we find out the there are situations where the link would be useful but off by default, we could add another hint to specify the default value. But for now there is no need to have them for all Vector2.
- I think it might depend on how it looks like, but removing the button will make the x/y components not aligned from one property to another. As a consequence it would be interesting to try moving it the right, to see if it looks better.

As a side note (that was not discussed in the PR meeting), do you think you could make the same changes for Vector3/Vector3i too?

KoBeWi commented:
> This is actually possible already, see ADD_LINKED_PROPERTY in object.h.

This is for dependent properties, it doesn't allow for conditional linking (like with the link button).

> As a side note (that was not discussed in the PR meeting), do you think you could make the same changes for Vector3/Vector3i too?

Yes, I was already planning to do it, just waited for the feature approval before putting more work.

KoBeWi commented:
Ok, reworked a bit. Moved the icon to the right and made the button hidden without the link hint:
![image](https://user-images.githubusercontent.com/2223172/161649458-ca03b58d-4058-45fe-9c6d-9a292f46b7fa.png)
I'll now add it to Vector3.

KoBeWi committed:
Add vector value linking

Co-authored-by: redlamp <244062+redlamp@users.noreply.github.com>

KoBeWi commented:
Ok added new icon contributed by @redlamp:
![](https://chat.godotengine.org/file-upload/vHskgMvBuxc8SrzBt/godot.windows.tools.64_QK73pLM4ox.gif)
The linked icon he provided was the same as Instance icon, so I reused Instance icon. It could be renamed to Linked maybe, but not in this PR.

akien-mga commented on editor/editor_properties.cpp:
```diff
@@ -1801,6 +1862,27 @@ void EditorPropertyVector3::_value_changed(double val, const String &p_name) {
+void EditorPropertyVector3::_update_ratio() {
+	linked->set_modulate(Color(1, 1, 1, linked->is_pressed() ? 1.0 : 0.5));
+
+	if (spin[0]->get_value() != 0 && spin[1]->get_value() != 0) {
+		ratio_yx = spin[1]->get_value() / spin[0]->get_value();
+		ratio_zx = spin[2]->get_value() / spin[0]->get_value();
+		ratio_xy = spin[0]->get_value() / spin[1]->get_value();
+		ratio_zy = spin[2]->get_value() / spin[1]->get_value();
+		ratio_xz = spin[0]->get_value() / spin[2]->get_value();
+		ratio_yz = spin[1]->get_value() / spin[2]->get_value();
+	} else {
+		ratio_yx = 1.0;
+		ratio_zx = 1.0;
+		ratio_xy = 1.0;
+		ratio_zy = 1.0;
+		ratio_xz = 1.0;
+		ratio_yz = 1.0;
+	}
```
`ratio_yx` is the same as `1.0 / ratio_xy`.
Maybe you could reduce the amount of member variables and computation needed by keeping only 3 and using `*` or `/` operations as appropriate?

KoBeWi replied:
I could, I even tried it at first, but IIRC there were some precision issues.

fire-forge commented:
> On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.

If you do add this in the future, perhaps it could be used to link the corner radius and border width properties of StyleBoxFlat. I find that I usually use the same values for all 4 sides/corners and having the link button there would make it easier to edit them all at once.

![image](https://user-images.githubusercontent.com/67974470/173196986-7b0043ac-5d98-4b95-8883-cbb47c0ff53c.png)

akien-mga reviewed (approved):
Should be good to merge once rebased.

akien-mga commented:
Thanks!

nathanfranke commented:
> > On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.
> 
> If you do add this in the future, perhaps it could be used to link the corner radius and border width properties of StyleBoxFlat. I find that I usually use the same values for all 4 sides/corners and having the link button there would make it easier to edit them all at once.
> 
> ![image](https://user-images.githubusercontent.com/67974470/173196986-7b0043ac-5d98-4b95-8883-cbb47c0ff53c.png)

Maybe these should be a single property with type [`Vector4`](https://github.com/godotengine/godot-proposals/issues/629) with a new property hint like `PROPERTY_HINT_SIDES` that changes `x`/`y`/`z`/`w` to `left`/`right`/`up`/`down` (or whatever the best order is).

timothyqiu commented:
Cherry-picked for 3.6
